sympy/core/function.py

```diff
@@ -348,6 +348,13 @@ def _eval_subs(self, old, new):
             old == self.func and len(self.args) in new.nargs):
             return new(*[i._subs(old, new) for i in self.args])
 
+    def doit(self, **hints):
+        deep = hints.get('deep', True)
```

Comment:
I don't want to add doit methods all over the codebase. It would be better to have `_eval_doit`.

Reply:
It's true, but there have been some problems.

Current `Basic.doit` is:
```python
def doit(self, **hints):
    if hints.get('deep', True):
        terms = [term.doit(**hints) if isinstance(term, Basic) else term for term in self.args]
        return self.func(*terms)
    else:
        return self
```

I was trying to change it like this and delete every `doit` of subclasses:
```python
def doit(self, **hints):
    if hasattr(self, '_eval_doit'):
        return self._eval_doit(**hints)
    if self.is_Atom:
        return self
    if hints.get('deep', True):
        terms = [term.doit(**hints) if isinstance(term, Basic) else term for term in self.args]
        return self.func(*terms, evaluate=True)
    else:
        return self.func(*self.args, evaluate=True)
```

However, this led to test failures because many classes don't accept `evaluate` option nor define their own `doit`.  
IMO, we should first make sure that every classes take `evaluate`. Only after then we can delete `doit`s, leave `Basic.doit` only, then make it refer to `_eval_doit`.

Reply:
Can you give some examples of your idea?